syweb/webclient/app/components/matrix/notification-service.js

```diff
@@ -48,23 +49,293 @@ angular.module('notificationService', ['matrixService'])
+                // Notify when another user joins
+                displayname = $filter("mUserDisplayName")(ev.state_key, ev.room_id);
+                message = displayname + " joined";
+            } else if (ev.state_key === matrixService.config().user_id  && "invite" === ev.content.membership) {
+                // notify when you are invited
+                message = displayname + " invited you to a room";
+            }
+        }
+        return message;
+    };
+
+    var matchingRuleForEventWithRulesets = function(ev, rulesets) {
+        if (!rulesets) return null;
+        if (ev.user_id == matrixService.config().user_id) return null;
+        angular.forEach(rulesets.device, function(devname, devrules) {
+            var matchingRule = matchingRuleFromKindSet(devrules, devname);
+            if (matchingRule) defer.resolve(matchingRule);
+        });
+        return matchingRuleFromKindSet(ev, rulesets.global);
+    };
+
+    var matchingRuleForEventNow = function(ev) {
+        // Returns a matching rule for this event with whatever rules
+        // we have cached right now (ie. it might be total rubbish)
+        return matchingRuleForEventWithRulesets(ev, rulesCache);
+    };
+
+    var shouldHighlightEventWithRule = function(ev, rule) {
+        var actionObj = actionListToActionsObject(rule.actions);
+        if (!actionObj.notify) return false;
+
+        if (actionObj.tweaks.highlight == undefined) {
+            // if it isn't specified, we highlight
+            return true;
+        }
+
+        return !!actionObj.tweaks.highlight;
+    };
+
+    var actionListToActionsObject = function(actionlist) {
+        var actionobj = { 'notify': false, 'tweaks': {} };
+        angular.forEach(actionlist, function(action) {
+            if (action === 'notify') {
+                actionobj.notify = true;
+            } else if (typeof action === 'object') {
+                if (action.value == undefined) action.value = true;
+                actionobj.tweaks[action.set_tweak] = action.value;
+            }
+        });
+        return actionobj;
+    };
+
+    var showNotification = function(title, body, icon, onclick, tag, audioNotification) {
+        if (!tag) {
+            tag = "matrix";
+        }
+        var notification = new window.Notification(
+            title,
+            {
+                "body": body,
+                "icon": icon,
+                "tag": tag
+            }
+        );
+
+        if (onclick) {
+            notification.onclick = onclick;
+        }
+        
+        var audioClip;
+        
+        if (audioNotification) {
+            audioClip = playAudio(audioNotification);
+        }
+
+        $timeout(function() {
+            notification.close();
+            if (audioClip) {
+                audioClip.pause();
+            }
+        }, 5 * 1000);
+    };
     
-    return {
+
+    var nfsvc = {
```

Comment:
Please don't abbreviate the final variable like this: it took me a while to understand that this was the variable that would be returned from the factory. It would be better to call it either ``notificationService`` or ``self``.